log4j-api/src/main/java/org/apache/logging/log4j/internal/DefaultLogBuilder.java

```diff
@@ -0,0 +1,179 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache license, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the license for the specific language governing permissions and
+ * limitations under the license.
+ */
+package org.apache.logging.log4j.internal;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.Marker;
+import org.apache.logging.log4j.message.Message;
+import org.apache.logging.log4j.message.SimpleMessage;
+import org.apache.logging.log4j.status.StatusLogger;
+import org.apache.logging.log4j.util.StackLocatorUtil;
+import org.apache.logging.log4j.util.Supplier;
+
+
+/**
+ * Collects data for a log event and then logs it.
+ */
+public class DefaultLogBuilder implements LogBuilder {
+    private static Message EMPTY_MESSAGE = new SimpleMessage("");
+    private static final String FQCN = DefaultLogBuilder.class.getName();
+    private static final Logger LOGGER = StatusLogger.getLogger();
+
+    private final Logger logger;
+    private Level level;
+    private Marker marker;
+    private Throwable throwable;
+    private StackTraceElement location;
+    private Object object;
+    private Message msg;
+    private String textMessage;
+    private Supplier<Message> supplier;
+    private Object[] parameters;
+    private volatile boolean inUse;
+    private long threadId;
+
+    public DefaultLogBuilder(Logger logger) {
+        this.logger = logger;
+        this.inUse = false;
+        this.threadId = Thread.currentThread().getId();
+    }
+
+    public LogBuilder setLevel(Level level) {
+        this.inUse = true;
+        this.level = level;
+        this.marker = null;
+        this.throwable = null;
+        this.location = null;
+        this.object = null;
+        this.msg = null;
+        this.textMessage = null;
+        this.supplier = null;
+        this.parameters = null;
+        return this;
```

Comment:
I believe the `setLevel` method should be moved to `Logger` (and renamed to `atLevel`), and removed from the `LogBuilder` class.

That will allow all of the remaining methods to skip processing if the log level is not applicable. That is one of the value propositions of [fluent logging APIs like this](https://google.github.io/flogger/benefits).

We lose that benefit if we allow users to change the log level later in the call chain: for example, this usage is possible:

```java
// assume log level is WARN
logger.trace().withMessage("This {} should {} be {} ignored {}")
    .withParameters("statement", "immediately", "completely", "for better performance")
    .setLevel(Level.TRACE) // oops, we cannot ignore the above after all
    .log();
```

Reply:
setLevel is dangerous and is not intended to be called by end users as it provides no value to them. As such it is not part of the LogBuilder interface and only exists in DefaultLogBuilder. As your example above shows, it would be nonsensical to create the LogBuilder by calling the trace method and then call setLevel() to set the level to something else. The main purpose of setLevel is to clear the LogBuilder for reuse.